### PR TITLE
Introduce wildcardhostname field

### LINKED ISSUE #949: HTTPRequestRedirectFilter allows for wildcard hostnames
**What happened**:

Looking through the `HTTPRequestRedirectFilter` spec, the `Hostname` used has a default kubebuilder validation regex which would allow a user to specify a redirect of `*.foo.com` which might not make sense as a redirect hostname. 

Controllers could catch this and set a condition, we could update the field to use a new `hostname` type which removes the ability to have a wildcard in the value, or maybe there's a use-case I'm not thinking of that folks would want this? 

**What you expected to happen**:

Hostnames with wildcards would be rejected. 

**How to reproduce it (as minimally and precisely as possible)**:

Here's a sample HTTPRoute:

```yaml
kind: HTTPRoute
apiVersion: gateway.networking.k8s.io/v1alpha2
metadata:
  name: kuard
  namespace: projectcontour
  labels:
    app: kuard
spec:
  parentRefs:
  - group: gateway.networking.k8s.io
    kind: Gateway
    name: contour
  hostnames:
  - "local.projectcontour.io"
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /
    filters:
      - requestRedirect:
          hostname: "*.stevesloka.com"
          scheme: https
          statusCode: 302
        type: RequestRedirect
    backendRefs:
    - kind: Service
      name: kuard
      port: 80
```

Right now Contour configures Envoy: 
```
$ curl -i local.projectcontour.io                                                                                                                                 
HTTP/1.1 302 Found
location: https://*.stevesloka.com/
vary: Accept-Encoding
date: Tue, 30 Nov 2021 17:37:35 GMT
server: envoy
content-length: 0
```

But doesn't route properly:
image

### PR DESCRIPTION
**Pull request by stevesloka**
**What type of PR is this?**
/kind bug

Optionally add one or more of the following kinds if applicable:
/kind api-change

**What this PR does / why we need it**:

Update the HTTPRequestRedirectFilter.Hostname to use a Hostname type which validates the value cannot
contain wildcard characters. Updates all other references which used Hostname to use WildcardHostname
which does allow for wildcard characters.

Signed-off-by: Steve Sloka <slokas@vmware.com>

**Which issue(s) this PR fixes**:
Fixes #949

**Does this PR introduce a user-facing change?**:
```release-note
Introduced WildcardHostname which allows for wildcard characters in relevant Hostname values. 
```

### CODE REVIEW

**Comment by hbagdi** [comment]
Can we add a test case  'wildcard hostname on the redirect filter fails'? You can do so by adding it to hack/invalid-examples.

File: apis/v1alpha2/gateway_types.go
Comment (jpeach) [inline apis/v1alpha2/gateway_types.go:177]: To me, it seems like it would be cleaner to change the type in the filter to `string` and add the custom  validation there. That's a smaller change and lets us keep the (IMHO nicer) `Hostname` typedef for the primary use case.

Comment (stevesloka) [reply apis/v1alpha2/gateway_types.go:177]: I agree about the smaller change @jpeach might be easier in the end and less user-breaking. 

@robscott does that work for you? I was originally going off of your suggestion of changing to a `WildcardHostname` as in the current commit.

Comment (robscott) [reply apis/v1alpha2/gateway_types.go:177]: We discussed this at the community meeting yesterday, meant to follow up sooner. I think at least @youngnick and I prefer using custom types like `Hostname` and `WildcardHostname` over a string so they're simpler to reuse in the future.

**Comment by stevesloka** [comment]
@hbagdi I added an invalid test case for the redirect hostname.

File: apis/v1alpha2/gateway_types.go
Comment (jpeach) [reply apis/v1alpha2/gateway_types.go:177]: In this case, however, it's one place that has custom behavior, so introducing a new type for that seems like a lot. If there are other hostname fields that can't be wildcards, I'd prefer to add the type at that time.

Continuing to rathole on this, what to people think of the type name `PreciseHostname` for a hostname that cannot contain a wildcard label? That's the terminology we use in description, so maybe it makes sense as a type too?

Comment (jpeach) [reply apis/v1alpha2/gateway_types.go:177]: I'm sorry to keep bikeshedding this ... `WildcardHostname` implies the name is a wildcard, but it's not. It might be wildcard or precise, so I don't think it's a good choice to replace `Hostname` (which I think is fine).

Comment (stevesloka) [reply apis/v1alpha2/gateway_types.go:177]: I agree with @jpeach that `Wilcardhostname` does imply it's "only" a wildcard. The `PreciseHostname` might be a good route to use for the redirect field and leave `Hostname` as-is.

Comment (robscott) [reply apis/v1alpha2/gateway_types.go:177]: 👍 that works for me.

Comment (jpeach) [reply apis/v1alpha2/gateway_types.go:177]: Me too, thanks @stevesloka :)

Comment (youngnick) [reply apis/v1alpha2/gateway_types.go:177]: I like the use of `PreciseHostname` as well, that helps to cement that usage as a way of distinguishing them in the code as well as the docs.

**Comment by youngnick** [comment]
LGTM once the field name changes are done as discussed.

**Comment by robscott** [comment]
@stevesloka I was about to release 0.4.1, but realized this change also probably belongs in that release. Agree with @youngnick that this PR LGTM once that final name change gets in.

**Comment by stevesloka** [comment]
@robscott sorry for loosing track of this one, I've updated the PR with the PreciseHostname. Thanks!

File: apis/v1alpha2/shared_types.go
Comment (jpeach) [inline apis/v1alpha2/shared_types.go:262]:
```suggestion
// the RFC 1123 definition of a hostname except that numeric IP addresses are not allowed.
```

Comment (stevesloka) [reply apis/v1alpha2/shared_types.go]: 👍

**Review by robscott — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Review by jpeach — APPROVED** [review]
/lgtm

**Comment by robscott** [comment]
Woops, we missed the hold on this.

/hold cancel

**Pull request merged**
